File: go/539.go
package main

import (
	"math"
	"slices"
	"strconv"
	"strings"
)

func getMinutes(str string) int {
	tokens := strings.Split(str, ":")

	hour, _ := strconv.Atoi(tokens[0])
	hour *= 60

	minutes, _ := strconv.Atoi(tokens[1])

	return minutes + hour
}

func findMinDifference(timePoints []string) int {
	result := math.MaxInt32

	m := []int{}
	for _, p := range timePoints {
		m = append(m, getMinutes(p))
	}

	slices.Sort(m)

	for i := 1; i < len(timePoints); i++ {
		result = min(result, m[i]-m[i-1])
	}

	return min(result, 24*60-m[len(m)-1]+m[0])
}

File: zig/539.zig
const std = @import("std");

fn getMinutes(str: []const u8) !i32 {
    var tokens = [2][]const u8{ "", "" };
    var it = std.mem.splitScalar(u8, str, ':');

    var idx: usize = 0;
    while (it.next()) |v| {
        tokens[idx] = v;
        idx += 1;
    }

    var hour = try std.fmt.parseInt(i32, tokens[0], 10);
    hour *= 60;

    const minutes = try std.fmt.parseInt(i32, tokens[1], 10);

    return minutes + hour;
}

fn solution(allocator: std.mem.Allocator, timeStrings: []const []const u8) !i32 {
    var temp = try allocator.alloc(i32, timeStrings.len);
    defer allocator.free(temp);

    for (timeStrings, 0..) |s, i| {
        temp[i] = try getMinutes(s);
    }

    std.mem.sort(i32, temp, {}, std.sort.asc(i32));

    var result = temp[1] - temp[0];

    for (2..temp.len) |idx| {
        result = @min(result, temp[idx] - temp[idx - 1]);
    }

    return @min(result, 24 * 60 - temp[temp.len - 1] + temp[0]);
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const allocator = std.testing.allocator;

    const input = [_][]const u8{ "23:59", "00:00" };

    const expected = 1;
    const actual = try solution(allocator, &input);

    try expect(expected, actual);
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const input = [_][]const u8{ "00:00", "23:59", "00:00" };

    const expected = 0;
    const actual = try solution(allocator, &input);

    try expect(expected, actual);
}
